<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ssRNA</th>
      <th>GEO_Accession</th>
      <th>BioProject_Accession</th>
      <th>Method</th>
      <th>Genome</th>
      <th>Article_DOI</th>
      <th>Author</th>
      <th>Pub_year</th>
      <th>Available_peaks_file</th>
      <th>Notes</th>
      <th>processed_RNABS_v</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>NEAT1</td>
      <td>GSM3417018</td>
      <td>SAMN10177340</td>
      <td>ASO_CaptureSeq, CaptureSeq</td>
      <td>Human</td>
      <td>10.1093/nar/gky1305</td>
      <td>Cetin</td>
      <td>2019</td>
      <td>https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM3417018&amp;format=file&amp;file=GSM3417018%5FCaptureSeq%5FNEAT1%5FTFR1%5Fpeaks%2Ebed%2Egz</td>
      <td>ASO-capture seq</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>1</th>
      <td>AL109615.3</td>
      <td>GSE136185</td>
      <td>PRJNA561525</td>
      <td>CHART-seq</td>
      <td>Human</td>
      <td>10.1158/0008-5472.CAN-20-2612.</td>
      <td>Zagorac</td>
      <td>2021</td>
      <td>no</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2</th>
      <td>MALAT1</td>
      <td>GSE58444</td>
      <td>PRJNA252626</td>
      <td>CHART-seq</td>
      <td>Human</td>
      <td>10.1016/j.molcel.2014.07.012</td>
      <td>West</td>
      <td>2014</td>
      <td>https://ars.els-cdn.com/content/image/1-s2.0-S1097276514006054-mmc2.xlsx</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>3</th>
      <td>NEAT1</td>
      <td>GSE58444</td>
      <td>PRJNA252626</td>
      <td>CHART-seq</td>
      <td>Human</td>
      <td>10.1016/j.molcel.2014.07.012</td>
      <td>West</td>
      <td>2014</td>
      <td>https://ars.els-cdn.com/content/image/1-s2.0-S1097276514006054-mmc2.xlsx</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>4</th>
      <td>1200007C13Rik</td>
      <td>GSE142169</td>
      <td>PRJNA596185</td>
      <td>ChIRP-seq</td>
      <td>Mouse</td>
      <td>10.1038/s41467-020-20207</td>
      <td>Chang</td>
      <td>2020</td>
      <td>https://static-content.springer.com/esm/art%3A10.1038%2Fs41467-020-20207-y/MediaObjects/41467_2020_20207_MOESM5_ESM.xlsx;https://static-content.springer.com/esm/art%3A10.1038%2Fs41467-020-20207-y/MediaObjects/41467_2020_20207_MOESM6_ESM.xlsx</td>
      <td>ChIRP- seq of MaTAR25 in mammary tumor cells to identify MaTAR25 target genes; the number of peaks provided by the authors is equal to the maximum number of rows in xls files</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>5</th>
      <td>7SK</td>
      <td>GSE69141</td>
      <td>PRJNA284635</td>
      <td>ChIRP-seq</td>
      <td>Human</td>
      <td>10.1038/nsmb.3176</td>
      <td>Flynn</td>
      <td>2016</td>
      <td>https://static-content.springer.com/esm/art%3A10.1038%2Fnsmb.3176/MediaObjects/41594_2016_BFnsmb3176_MOESM31_ESM.xlsx</td>
      <td>snRNP</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>6</th>
      <td>9530018H14Rik\n</td>
      <td>GSE70986</td>
      <td>PRJNA290073</td>
      <td>ChIRP-seq</td>
      <td>Mouse</td>
      <td>10.1371/journal.pgen.1005680</td>
      <td>Carlson</td>
      <td>2015</td>
      <td>https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE70986&amp;format=file&amp;file=GSE70986%5FLimb%2Dpeaks%2Ebed%2Egz;https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE70986&amp;format=file&amp;file=GSE70986%5Fglial%2Dpeaks%2Ebed%2Egz</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>7</th>
      <td>allRNAs</td>
      <td>GSE97131</td>
      <td>PRJNA380782</td>
      <td>ChAR-seq</td>
      <td>Drosophila</td>
      <td>10.7554/eLife.27024.001</td>
      <td>Bell</td>
      <td>2018</td>
      <td>NaN</td>
      <td>(2018)ChAR-seq, a proximity ligation and sequencing method that both identifies chromatin-associated RNAs and maps them to genomic loci</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>8</th>
      <td>allRNAs</td>
      <td>GSE82312</td>
      <td>PRJNA324602</td>
      <td>GRID-seq</td>
      <td>Mouse</td>
      <td>10.1038/nbt.3968</td>
      <td>Li</td>
      <td>2017</td>
      <td>NaN</td>
      <td>(2017)Global RNA-chromatin interactions in 3 cell typer of 3 species</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>9</th>
      <td>allRNAs</td>
      <td>GSE82312</td>
      <td>PRJNA324602</td>
      <td>GRID-seq</td>
      <td>Human</td>
      <td>10.1038/nbt.3968</td>
      <td>Li</td>
      <td>2017</td>
      <td>NaN</td>
      <td>(2017)Global RNA-chromatin interactions in 3 cell typer of 3 species</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>10</th>
      <td>allRNAs</td>
      <td>GSE82312</td>
      <td>PRJNA324602</td>
      <td>GRID-seq</td>
      <td>Drosophila</td>
      <td>10.1038/nbt.3968</td>
      <td>Li</td>
      <td>2017</td>
      <td>NaN</td>
      <td>(2017)Global RNA-chromatin interactions in 3 cell typer of 3 species</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>11</th>
      <td>allRNAs</td>
      <td>GSE92345</td>
      <td>PRJNA357269</td>
      <td>MARGI-seq</td>
      <td>Human</td>
      <td>10.1016/j.cub.2017.01.011</td>
      <td>Sridhar</td>
      <td>2017</td>
      <td>NaN</td>
      <td>(2017)pxMARGI is designed to not differentiate passive and direct interactions, diMARGI is aimed to reveal protein tethered interactions</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>12</th>
      <td>allRNAs</td>
      <td>GSE132192</td>
      <td>PRJNA546280</td>
      <td>RADICL-seq</td>
      <td>Mouse</td>
      <td>10.1038/s41467-020-14337-6</td>
      <td>Bonetti</td>
      <td>2020</td>
      <td>NaN</td>
      <td>(2020)RADICL-seq is a proximity ligation-based methodology that reduces the bias for nascent transcription, while increasing genomic coverage and unique mapping rate efficiency compared with existing methods</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>13</th>
      <td>AC018781.1</td>
      <td>GSE104479</td>
      <td>PRJNA412810</td>
      <td>ChIRP-seq</td>
      <td>Human</td>
      <td>https://doi.org/10.1002/art.41200</td>
      <td>Fan</td>
      <td>2020</td>
      <td>https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE104479&amp;format=file</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>14</th>
      <td>AC087482.1</td>
      <td>GSE58989</td>
      <td>PRJNA254057</td>
      <td>ChIRP-seq</td>
      <td>Human</td>
      <td>10.1093/hmg/ddv343</td>
      <td>Merry</td>
      <td>2015</td>
      <td>https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM1854427&amp;format=file&amp;file=GSM1854427%5FDACOR1%5FChIRP%5Fanalysis%2Etxt%2Egz</td>
      <td>Equal amounts of nuclear lysates were incubated with either DACOR1 specific or non-specific DNA probes modified with a TEG linker and Biotin at their 5â€™ ends</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>15</th>
      <td>AK156552.1</td>
      <td>GSE73805</td>
      <td>PRJNA297950</td>
      <td>ChIRP-seq</td>
      <td>Mouse</td>
      <td>10.1016/j.celrep.2017.11.045</td>
      <td>Wang</td>
      <td>2018</td>
      <td>https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE73805&amp;format=file&amp;file=GSE73805%5FPanct1%5FWT%5FOdd%5Fand%5FEven%2Ebed%2Egz</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>16</th>
      <td>Bloodlinc</td>
      <td>GSE97119</td>
      <td>PRJNA380759</td>
      <td>ChIRP-seq</td>
      <td>Mouse</td>
      <td>10.1016/j.celrep.2017.05.082</td>
      <td>Alvarez-Dominguez</td>
      <td>2017</td>
      <td>https://ftp.ncbi.nlm.nih.gov/geo/series/GSE97nnn/GSE97119/suppl/GSE97119_EC7_consensus_aboveThresholds_peaks.bed.gz</td>
      <td>alncRNA-EC7/Bloodlinc, is transcribed from a super-enhancer of the erythroid membrane transporter SLC4A1/BAND3 but diffuses beyond this site</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>17</th>
      <td>CDKN2B-AS1</td>
      <td>GSE162221</td>
      <td>PRJNA681143</td>
      <td>ChIRP-seq</td>
      <td>Human</td>
      <td>10.1093/nar/gkab245</td>
      <td>Alfeghaly</td>
      <td>2021</td>
      <td>https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE162221&amp;format=file&amp;file=GSE162221%5Fintersect%5FANRIL%5Fpeaks%5Frep1%5Frep2%5Fhighscore15%5Fblacklisted%5Fregions%2Ebed%2Egz</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>18</th>
      <td>CPEB2-DT</td>
      <td>GSE156121</td>
      <td>PRJNA656806</td>
      <td>ChIRP-seq</td>
      <td>Human</td>
      <td>10.1172/jci.insight.143289</td>
      <td>Reddy</td>
      <td>2021</td>
      <td>https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE156121&amp;format=file&amp;file=GSE156121%5FTHP1%2DDRAIR%5Frep1%2D1e%2D5%5Fpeaks%2Etxt%2Egz;https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE156121&amp;format=file&amp;file=GSE156121%5FTHP1%2DDRAIR%5Frep2%2D1e%2D5%5Fpeaks%2Etxt%2Egz</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>19</th>
      <td>DT926623.1</td>
      <td>GSE66819</td>
      <td>PRJNA278066</td>
      <td>ChIRP-seq</td>
      <td>Mouse</td>
      <td>10.1016/j.stem.2015.02.002</td>
      <td>Luo</td>
      <td>2015</td>
      <td>https://ftp.ncbi.nlm.nih.gov/geo/series/GSE66nnn/GSE66819/suppl/GSE66819_final.consensus.narrowpeak.bed.gz</td>
      <td>LincHSC-2 binding sites were interrogated in HPC5 mouse bone marrow progenitor cells by ChIRP-seq</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>20</th>
      <td>Eprn</td>
      <td>GSE147208</td>
      <td>PRJNA613423</td>
      <td>ChIRP-seq</td>
      <td>Mouse</td>
      <td>10.1093/nar/gkaa628</td>
      <td>Zapparoli</td>
      <td>2020</td>
      <td>no</td>
      <td>EPR (BC030870) ChIRP-Seq analysis in NMuMG mouse mammary gland</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>21</th>
      <td>ERV-9 assoc. lncRNAs</td>
      <td>GSE93281</td>
      <td>PRJNA360462</td>
      <td>ChIRP-seq</td>
      <td>Mouse</td>
      <td>10.1093/nar/gkx055</td>
      <td>Hu</td>
      <td>2017</td>
      <td>no</td>
      <td>ERV-9 LTR lncRNAs</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>22</th>
      <td>ERV-9 assoc. lncRNAs</td>
      <td>GSE93281</td>
      <td>PRJNA360462</td>
      <td>ChIRP-seq</td>
      <td>Human</td>
      <td>10.1093/nar/gkx055</td>
      <td>Hu</td>
      <td>2017</td>
      <td>no</td>
      <td>ERV-9 LTR lncRNAs</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>23</th>
      <td>FA2H-2-MLKL</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>ChIRP-seq</td>
      <td>Human</td>
      <td>10.1038/s41418-018-0235-z</td>
      <td>Guo</td>
      <td>2019</td>
      <td>no</td>
      <td>no data available</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>24</th>
      <td>Gm4665</td>
      <td>GSE145189</td>
      <td>PRJNA606277</td>
      <td>ChIRP-seq</td>
      <td>Mouse</td>
      <td>10.1101/gr.264333.120</td>
      <td>Li</td>
      <td>2021</td>
      <td>https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE145189&amp;format=file&amp;file=GSE145189%5FMACS2%5FLncRNA%5FInput%2Etxt%2Egz</td>
      <td>Analysis of the retrieved chromatin fragments by deep sequencing (ChIRP-seq) revealed Gm4665 occupancy sites were genome wide, specifically abundant enrich in intronic regions</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>25</th>
      <td>HAND2-AS1</td>
      <td>GSE126122</td>
      <td>PRJNA521057</td>
      <td>ChIRP-seq</td>
      <td>Human</td>
      <td>10.15252/embj.2018101110</td>
      <td>wang</td>
      <td>2019</td>
      <td>no</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>26</th>
      <td>HOTAIR</td>
      <td>GSE31332</td>
      <td>PRJNA146003</td>
      <td>ChIRP-seq</td>
      <td>Human</td>
      <td>10.1016/j.molcel.2011.08.027</td>
      <td>CiChu</td>
      <td>2011</td>
      <td>https://ftp.ncbi.nlm.nih.gov/geo/series/GSE31nnn/GSE31332/suppl/GSE31332%5Fhotair%5Foe%5Fpeaks%2Ebed%2Egz</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>27</th>
      <td>HOTTIP</td>
      <td>GSE114981</td>
      <td>PRJNA473439</td>
      <td>ChIRP-seq</td>
      <td>Human</td>
      <td>10.1016/j.ccell.2019.10.011</td>
      <td>Luo</td>
      <td>2019</td>
      <td>https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM3161929&amp;format=file&amp;file=GSM3161929%5FWT%2DCHIRP%2Dseq%5Fpeaks%2Ebed%2Egz</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>28</th>
      <td>Lin28a</td>
      <td>GSE90574</td>
      <td>PRJNA355091</td>
      <td>ChIRP-seq</td>
      <td>Mouse</td>
      <td>10.7554/eLife.23468</td>
      <td>Li</td>
      <td>2017</td>
      <td>https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE90574&amp;format=file</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>29</th>
      <td>Linc-ASEN</td>
      <td>GSE128396</td>
      <td>PRJNA527315</td>
      <td>ChIRP-seq</td>
      <td>Human</td>
      <td>10.1038/s41418-019-0467-6</td>
      <td>Lee</td>
      <td>2020</td>
      <td>https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM3673641&amp;format=file&amp;file=GSM3673641%5F713%5F2%5Fp01%5Fpeaks%2Ebed%2Egz</td>
      <td>no input or negative control</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>30</th>
      <td>lncDUSP</td>
      <td>GSE112602</td>
      <td>PRJNA448439</td>
      <td>ChIRP-seq</td>
      <td>Human</td>
      <td>10.1038/s41598-018-25530-5</td>
      <td>Forrest</td>
      <td>2018</td>
      <td>https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM3073889&amp;format=file&amp;file=GSM3073889%5FlincDUSP%5FChIRP%2Dseq%5Fpeaks%2Exlsx</td>
      <td>Small number of reads in fastqs (3543)</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>31</th>
      <td>lncSmad7</td>
      <td>GSE174636</td>
      <td>PRJNA730846</td>
      <td>ChIRP-seq</td>
      <td>Mouse</td>
      <td>doi.org/10.1093/nar/gkac083</td>
      <td>Maldotti</td>
      <td>2022</td>
      <td>no (direct target genes only)</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>32</th>
      <td>Meg3</td>
      <td>GSE99798</td>
      <td>PRJNA389664</td>
      <td>ChIRP-seq</td>
      <td>Mouse</td>
      <td>10.1128/MCB.00278-17</td>
      <td>Iyer</td>
      <td>2017</td>
      <td>https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE99798&amp;format=file&amp;file=GSE99798%5FM5%5Funique%5FNGSA%2Ebed%2Egz</td>
      <td>Triplex evidence</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>33</th>
      <td>MIR503HG</td>
      <td>PRJEB34589</td>
      <td>NaN</td>
      <td>ChIRP-seq</td>
      <td>Human</td>
      <td>10.1172/JCI135439</td>
      <td>Pachera</td>
      <td>2020</td>
      <td>no</td>
      <td>ChIRP-Seq in the fibroblast cell line BJ5TA after 6 hours of TGF-Î² stimulation.</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>34</th>
      <td>NFKBMARL-1</td>
      <td>GSE167504</td>
      <td>PRJNA704835</td>
      <td>ChIRP-seq</td>
      <td>Human</td>
      <td>10.1038/s43587-021-00056-0</td>
      <td>Cai</td>
      <td>2021</td>
      <td>https://static-content.springer.com/esm/art%3A10.1038%2Fs43587-021-00056-0/MediaObjects/43587_2021_56_MOESM6_ESM.xlsx</td>
      <td>non si distinguono controllo positivo, negativo o input</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>35</th>
      <td>NRAD1</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>ChIRP-seq</td>
      <td>Human</td>
      <td>10.1038/s41418-019-0362-1</td>
      <td>Vidovic</td>
      <td>2020</td>
      <td>no</td>
      <td>no data available</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>36</th>
      <td>PCGEM1</td>
      <td>GSE47804</td>
      <td>PRJNA207830</td>
      <td>ChIRP-seq</td>
      <td>Human</td>
      <td>10.1038/nature12451</td>
      <td>Yang</td>
      <td>2013</td>
      <td>https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE47804&amp;format=file</td>
      <td>no input or negative control</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>37</th>
      <td>Rmrp</td>
      <td>GSE73912</td>
      <td>PRJNA298543</td>
      <td>ChIRP-seq</td>
      <td>Mouse</td>
      <td>10.1038/nature16193</td>
      <td>Huang</td>
      <td>2015</td>
      <td>https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE73912&amp;format=file&amp;file=GSE73912%5FRMRP%5FChIRPseq%5FDDX5ko%5FoverlapPeaks%2Ebed%2Egz;https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE73912&amp;format=file&amp;file=GSE73912%5FRMRP%5FChIRPseq%5FWT%5FoverlapPeaks%2Ebed%2Egz</td>
      <td>Retracted article</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>38</th>
      <td>Rn7sk</td>
      <td>GSE69141</td>
      <td>PRJNA284635</td>
      <td>ChIRP-seq</td>
      <td>Mouse</td>
      <td>10.1038/nsmb.3176</td>
      <td>Flynn</td>
      <td>2016</td>
      <td>https://static-content.springer.com/esm/art%3A10.1038%2Fnsmb.3176/MediaObjects/41594_2016_BFnsmb3176_MOESM31_ESM.xlsx</td>
      <td>snRNA</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>39</th>
      <td>SNHG11</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>ChIRP-seq</td>
      <td>Human</td>
      <td>10.1038/s41388-020-01512-8</td>
      <td>Xu</td>
      <td>2020</td>
      <td>https://static-content.springer.com/esm/art%3A10.1038%2Fs41388-020-01512-8/MediaObjects/41388_2020_1512_MOESM2_ESM.xls</td>
      <td>no data available</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>40</th>
      <td>SRA1</td>
      <td>GSE153004</td>
      <td>PRJNA641150</td>
      <td>ChIRP-seq</td>
      <td>Human</td>
      <td>10.3389/fgene.2020.00850</td>
      <td>Sawaengdee</td>
      <td>2020</td>
      <td>https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM4632919&amp;format=file&amp;file=GSM4632919%5FK562%5FSRA%5F1%5Fp0001%5Fpeaks%2Ebed%2Egz;https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM4632920&amp;format=file&amp;file=GSM4632920%5FK562%5FSRA%5F2%5Fp0001%5Fpeaks%2Ebed%2Egz;https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM4632921&amp;format=file&amp;file=GSM4632921%5FK562%5FSRA%5F3%5Fp0001%5Fpeaks%2Ebed%2Egz</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>41</th>
      <td>SRA1</td>
      <td>GSE58641</td>
      <td>PRJNA253062</td>
      <td>ChIRP-seq</td>
      <td>Human</td>
      <td>10.1371/journal.pgen.1005615</td>
      <td>Wongtrakoongate</td>
      <td>2015</td>
      <td>https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM1415922&amp;format=file&amp;file=GSM1415922%5FNTERA2%5FSRA%2Ebed%2Egz</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>42</th>
      <td>TERC</td>
      <td>GSE31332</td>
      <td>PRJNA146003</td>
      <td>ChIRP-seq</td>
      <td>Human</td>
      <td>10.1016/j.molcel.2011.08.027</td>
      <td>CiChu</td>
      <td>2011</td>
      <td>https://ftp.ncbi.nlm.nih.gov/geo/series/GSE31nnn/GSE31332/suppl/GSE31332%5Fterc%5Fpeaks%2Ebed%2Egz</td>
      <td>Triplex evidence (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6735767/)</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>43</th>
      <td>Tug1</td>
      <td>GSE77493</td>
      <td>PRJNA310632</td>
      <td>ChIRP-seq</td>
      <td>Mouse</td>
      <td>10.1172/JCI87927</td>
      <td>Long</td>
      <td>2016</td>
      <td>https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE77493&amp;format=file</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>44</th>
      <td>Z85996.1</td>
      <td>GSE76418</td>
      <td>PRJNA307254</td>
      <td>ChIRP-seq</td>
      <td>Human</td>
      <td>10.1038/ng.3673</td>
      <td>Schmitt</td>
      <td>2016</td>
      <td>https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM2011201&amp;format=file&amp;file=GSM2011201%5FAS1%5F1%5FDINO%5FChIRP%5FEVEN%2EnarrowPeak%2Egz;https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM2011202&amp;format=file&amp;file=GSM2011202%5FAS1%5F2%5FDINO%5FChIRP%5FODD%2EnarrowPeak%2Egz</td>
      <td>Examine the chormatin localization of DINO in DNA damaged human U2OS cells</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>45</th>
      <td>MEG3</td>
      <td>PRJEB7307</td>
      <td>NaN</td>
      <td>ChOP-seq</td>
      <td>Human</td>
      <td>10.1038/ncomms8743</td>
      <td>Mondal</td>
      <td>2015</td>
      <td>https://static-content.springer.com/esm/art%3A10.1038%2Fncomms8743/MediaObjects/41467_2015_BFncomms8743_MOESM364_ESM.xls</td>
      <td>Triplex evidence</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>46</th>
      <td>Xist</td>
      <td>GSE137305</td>
      <td>PRJNA565139</td>
      <td>RAP-seq</td>
      <td>Mouse</td>
      <td>10.1126/science.1237973</td>
      <td>Engreitz</td>
      <td>2013</td>
      <td>NaN</td>
      <td>chrX only</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>